src/Pos/Update/Poll/Trans.hs

```diff
@@ -90,93 +88,56 @@ execPollT m = fmap snd . runPollT m
-               | Just res <- new -> purer res
-               | otherwise -> getBVState pv
-    getProposedBVs =
-        PollT $ do
-            new <- use pmNewBVsL
-            del <- use pmDelBVsL
-            underlying <- getProposedBVs
-            let filteredU = filter (not . flip HS.member del) underlying
-            return $ HM.keys new <> filteredU
+    getBVState pv = PollT $ MM.lookupM getBVState pv =<< use pmBVsL
+    getProposedBVs = PollT $ MM.keysM getProposedBVs =<< use pmBVsL
     getConfirmedBVStates =
-        PollT $ do
-            new <- use pmNewBVsL
-            del <- use pmDelBVsL
-            underlying <- getConfirmedBVStates
-            let filteredU = filter (not . flip HS.member del . fst) $ underlying
-            let filteredN = filter (bvsIsConfirmed . snd) $ HM.toList new
-            return $ filteredU <> filteredN
+        PollT $
+        filter (bvsIsConfirmed . snd) <$>
+        (MM.toListM getConfirmedBVStates =<< use pmBVsL)
     getAdoptedBVFull =
         PollT $ maybe getAdoptedBVFull pure =<< use pmAdoptedBVFullL
     getLastConfirmedSV appName =
-        PollT $ do
-            new <- use $ pmNewConfirmedL . at appName
-            del <- use $ pmDelConfirmedL . at appName
-            if | Just () <- del -> return Nothing
-               | Just res <- new -> purer res
-               | otherwise -> getLastConfirmedSV appName
-    hasActiveProposal appName = PollT $ do
-        new <- pmNewActivePropsIdx <$> get
-        del <- pmDelActivePropsIdx <$> get
-        if | appName `HM.member` del -> return False
-           | appName `HM.member` new -> return True
-           | otherwise -> hasActiveProposal appName
-    getProposal upId = PollT $ do
-        new <- pmNewActiveProps <$> get
-        del <- pmDelActiveProps <$> get
-        if | upId `HS.member` del -> return Nothing
-           | Just res <- HM.lookup upId new -> return (Just res)
-           | otherwise -> getProposal upId
+        PollT $ MM.lookupM getLastConfirmedSV appName =<< use pmConfirmedL
+    hasActiveProposal appName =
+        PollT $ do
+            new <- pmNewActivePropsIdx <$> get
+            del <- pmDelActivePropsIdx <$> get
+            if | appName `HM.member` del -> return False
+               | appName `HM.member` new -> return True
+               | otherwise -> hasActiveProposal appName
+    getProposal upId =
+        PollT $ MM.lookupM getProposal upId =<< use pmActivePropsL
     getConfirmedProposals =
-        PollT $ do
-            new <- use pmNewConfirmedPropsL
-            del <- use pmDelConfirmedPropsL
-            underlying <- getConfirmedProposals
-            let filtered =
-                    filter
-                        (not . flip HS.member del . cpsSoftwareVersion)
-                        underlying
-            return $ toList new ++ filtered
+        PollT $
+        MM.valuesM
+            (map (first cpsSoftwareVersion . join (,)) <$> getConfirmedProposals) =<<
```

Comment:
I was told that `\x -> (f x, x)` is bad and `map (first f . join (,))` is more readable. It am not sure about that but I just believed.